zio-json/shared/src/main/scala/zio/json/ast/JsonCursor.scala

```diff
@@ -1,58 +1,60 @@
 package zio.json.ast
 
-import scala.annotation.implicitNotFound
+sealed trait JsonCursor[From, To <: Json] { self =>
+  final def >>>[Next <: Json](that: JsonCursor[To, Next]) =
+    JsonCursor.AndThen(self, that)
 
-// TODO: Add another type parameter (From) to JsonCursor
-// TODO: Rename to JsonPath[From, To]???
-sealed trait JsonCursor[A <: Json] { self =>
-  // TODO: Rename to >>>
-  def ++[B <: Json](that: JsonCursor[B])(implicit drillDown: JsonCursor.DrillDown[A, B]): JsonCursor[B] = ???
+  def isArray: JsonCursor[Json, Json.Arr] = filterType(JsonType.Arr)
 
-  def isArray: JsonCursor[Json.Arr] = filterType(JsonType.Arr)
+  def isBool: JsonCursor[Json, Json.Bool] = filterType(JsonType.Bool)
 
-  def isBool: JsonCursor[Json.Bool] = filterType(JsonType.Bool)
+  def filterType[B <: Json](jsonType: JsonType[B]): JsonCursor[Json, B] =
+    JsonCursor.FilterType(self, jsonType)
 
-  def filterType[B <: Json](jsonType: JsonType[B]): JsonCursor[B] = JsonCursor.FilterType(self, jsonType)
+  def isNull: JsonCursor[Json, Json.Null] = filterType(JsonType.Null)
 
-  def isNull: JsonCursor[Json.Null] = filterType(JsonType.Null)
+  def isNumber: JsonCursor[Json, Json.Num] = filterType(JsonType.Num)
 
-  def isNumber: JsonCursor[Json.Num] = filterType(JsonType.Num)
+  def isObject: JsonCursor[Json, Json.Obj] = filterType(JsonType.Obj)
 
-  def isObject: JsonCursor[Json.Obj] = filterType(JsonType.Obj)
-
-  def isString: JsonCursor[Json.Str] = filterType(JsonType.Str)
+  def isString: JsonCursor[Json, Json.Str] = filterType(JsonType.Str)
 }
+
 object JsonCursor {
-  @implicitNotFound("Cannot drill down from ${A} into ${B}")
-  sealed trait DrillDown[A, B]
-  object DrillDown {
-    implicit def LeftIsJson[B]: DrillDown[Json, B]    = new DrillDown[Json, B] {}
-    implicit def LeftIsArr[B]: DrillDown[Json.Arr, B] = new DrillDown[Json.Arr, B] {}
-    implicit def LeftIsObj[B]: DrillDown[Json.Obj, B] = new DrillDown[Json.Obj, B] {}
+  implicit class JsonCursorObjOps[From](cursor: JsonCursor[From, Json.Obj]) {
```

Comment:
@jdegoes Could not define this on `JsonCursor` without casting, even though the compiler should know that the is sound, this is what it looked like:

```scala
  def field(field: String)(implicit ev: To =:= Json.Obj) =
   JsonCursor.DownField(self.asInstanceOf[JsonCursor[From, Json.Obj]], field
```

Reply:
It's because you need to map on the output type. 

In thinking about this, I believe that we should have:

```scala
sealed trait JsonCursor[-From, +To]
```

The use of variance will make composition nicer and eliminate some type casts.

Reply:
Makes sense, will revisit!

Reply:
Btw, I always recommend using `A <:< B` instead of `A =:= B` because it's more flexible.

Reply:
/cc @fsvehla

Reply:
👍🏻